ttnn/cpp/ttnn/tensor/vector_base.cpp

```diff
@@ -3,40 +3,66 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "vector_base.hpp"
+#include <stdexcept>
+#include "fmt/color.h"
+#include "tt_metal/common/assert.hpp"
 
 namespace tt::tt_metal {
 
 namespace {
-int32_t normalized_index(int32_t index, size_t container_size) {
-    int32_t size = static_cast<int32_t>(container_size);
 
-    if (index < 0) {
-        index += size;
+int32_t normalized_index(int32_t index, size_t original_size, size_t container_size) {
+    int32_t orig_size = static_cast<int32_t>(original_size);
+    int32_t full_size = static_cast<int32_t>(container_size);
+
+    int fixed_index = index;
+    if (fixed_index < 0) {
+        fixed_index += full_size;
+    } else {
+        fixed_index += full_size - orig_size;
     }
 
-    if (index < 0 || index >= size) {
-        throw std::out_of_range("SimpleShape index out of range.");
+    if (fixed_index < 0 || fixed_index >= full_size) {
+        TT_THROW("VectorBase[] index out of range. {} not in [{}, {})", index, -full_size, full_size);
     }
 
-    return index;
+    return fixed_index;
+}
 }
+
+void VectorBase::init() {
+    m_original_size = m_value.size();
+    const size_t min_internal_size = 4;
+    if(m_original_size < min_internal_size) {
+        std::vector<uint32_t> new_value(min_internal_size, 1);
+        std::copy(m_value.begin(), m_value.end(), new_value.begin() + (min_internal_size - m_original_size));
+        m_value = std::move(new_value);
```

Comment:
just curious, why not just reserve elements? probably i am missing something

```
if (m_original_size < min_internal_size) {
   m_value.reserve(min_internal_size);
}
```

Reply:
if you prefer your way, can we make same thing with `insert`?

Reply:
We need to insert elements into the beginning

Reply:
yep, is it a problem?

Reply:
Will use insert as discussed offline